pkg/api/types.go

```diff
@@ -2128,3 +2128,35 @@ type RangeAllocation struct {
 	// a single allocated address (the fifth bit on CIDR 10.0.0.0/8 is 10.0.0.4).
 	Data []byte `json:"data"`
 }
+
+// A ThirdPartyObject is a generic representation of a resource, it is used by add-ons and plugins to add new resource
+// types to the API.  It consists of one or more Versions of the api.
+type ThirdPartyObject struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty" description:"standard object metadata"`
+
+	Description string `json:"description,omitempty" description:"The description of this object"`
+
+	Versions []Version `json:"versions,omitempty" description:"The versions for this third party object"`
+}
+
+type ThirdPartyObjectList struct {
+	TypeMeta `json:",inline"`
+	ListMeta `json:"metadata,omitempty" description:"standard list metadata; see http://docs.k8s.io/api-conventions.md#metadata"`
+
+	Items []ThirdPartyObject `json:"items" description:"items is a list of schema objects"`
+}
+
+// A Version represents a single concrete version of an object model.
+type Version struct {
+	Name     string `json:"name,omitempty" description:"name of this version (e.g. 'v1')"`
+	APIGroup string `json:"apiGroup,omitempty" description:"The API group to add this object into, default 'experimental'"`
+}
+
+// An internal object, used for versioned storage in etcd.  Not exposed to the end user.
+type ThirdPartyObjectData struct {
```

Comment:
We probably need to figure out how it will interact with the types defined in pkg/runtime/types.go. I think we'll need the group X version X kind to be mapped to this type rather than degenerate to Unknown.

cc @smarterclayton